Allow to stop pre-parameters generator with a context

pdyraga opened the pull request:
Refs [#3161](https://github.com/keep-network/keep-core/issues/3161)

I am marking this PR as a draft hoping for an internal review process before I open a PR to `bnb-chain/tss-lib`.

Generating DKG pre-parameters is a computationally expensive operation. As `GeneratePreParams` documentation explains, it is recommended to do it out of band. Keep ECDSA (TBTC) client has a [pre-parameters](https://github.com/keep-network/keep-core/blob/main/pkg/tecdsa/dkg/params_pool.go) pool from which the DKG process pulls pre-parameters to do not have the entire group wait with kicking off DKG until pre-parameters are generated. This approach was sufficient for TBTC v1 when signing groups were 3-of-3. For TBTC v2, we need much larger signing groups (51-of-100) and for clients having multiple seats in the signing group, generating multiple pre-parameters in a row eliminates other processes run by the client from the access to the CPU.

To fix this problem, we plan to run pre-parameters generator only when no other processes are currently executed by the client. That is, when the client would normally be idle, it will keep generating pre-parameters, and when the client starts a protocol execution (e.g. DKG), pre-parameters generation should be stopped.

`GeneratePreParams` allows specifying a timeout which is really helpful but is not enough in this case. We want to stop the generator immediately when some protocol is about to start.

To achieve it, the code has been refactored to accept a context as a parameter of all long-running generator functions. The context can be canceled or can be constructed in such a way that it times out after a certain time. To achieve backward compatibility, `GeneratePreParams` signature does not change and a new function accepting the context `GeneratePreParamsWithContext` has been introduced.

pdyraga committed:
Allow to stop pre-parameters generator with a context

Generating DKG pre-parameters is a computationally expensive operation.
As `GeneratePreParams` documentation explains, it is recommended to do it out of
band. Keep ECDSA (TBTC) client has a pre-parameters pool from which the DKG
process pulls pre-parameters to do not have the entire group wait with kicking
off DKG until pre-parameters are generated. This approach was sufficient for
TBTC v1 when signing groups were 3-of-3. For TBTC v2, we need much larger
signing groups (51-of-100) and for clients having multiple seats in the signing
group, generating multiple pre-parameters in a row eliminates other processes
run by the client from the access to the CPU.

To fix this problem, we plan to run pre-parameters generator only when no other
processes are currently executed by the client. That is, when the client would
normally be idle, it will keep generating pre-parameters and when the client
starts a protocol execution (e.g. DKG), pre-parameters generation should be
stopped.

GeneratePreParams allows specifying a timeout which is really helpful but is not
enough in this case. We want to stop the generator immediately when some
protocol is about to start.

To achieve it, the code has been refactored to accept a context as a parameter
of all long-running generator functions. The context can be canceled or can be
constructed in such a way that it times out after a certain time. To achieve
backward compatibility, GeneratePreParams signature does not change and a new
function accepting the context GeneratePreParamsWithContext has been introduced.

lukasz-zimnoch commented on common/safe_prime.go:
```diff
@@ -121,7 +125,7 @@ var smallPrimesProduct = new(big.Int).SetUint64(16294579238595022365)
 // This function generates safe primes of at least 6 `bitLen`. For every
 // generated safe prime, the two most significant bits are always set to `1`
 // - we don't want the generated number to be too small.
-func GetRandomSafePrimesConcurrent(bitLen, numPrimes int, timeout time.Duration, concurrency int) ([]*GermainSafePrime, error) {
+func GetRandomSafePrimesConcurrent(ctx context.Context, bitLen, numPrimes int, concurrency int) ([]*GermainSafePrime, error) {
```
This function is public so this is a potential breaking change. Shouldn't we expose ` GetRandomSafePrimesConcurrentWithContext` for full backward compatibility?

pdyraga replied:
I was thinking about it but I am not sure if it makes sense to complicate the code. I think `common.GetRandomnSafePrimesConcurrent` and `paillier.GenerageKeyPair` are not something the code using `tss-lib` calls.

lukasz-zimnoch commented on crypto/paillier/paillier.go:
```diff
@@ -65,7 +65,7 @@ func init() {
 }
 
 // len is the length of the modulus (each prime = len / 2)
-func GenerateKeyPair(modulusBitLen int, timeout time.Duration, optionalConcurrency ...int) (privateKey *PrivateKey, publicKey *PublicKey, err error) {
```
This function is public so this is a potential breaking change. Shouldn't we expose ` GenerateKeyPairWithContext` for full backward compatibility?

lukasz-zimnoch reviewed (commented):
LGTM! Just two comments.

pdyraga committed:
Increased duration delta to 1s in prepare_test.go

10ms delta was too aggresive - it was fine for local dev environment on
Apple M1 but CI builds based on GitHub Actions kept failing. Increased
to 1s.

pdyraga commented:
Closing in favor of https://github.com/bnb-chain/tss-lib/pull/191.
